presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/RandomizeNullKeyInOuterJoin.java

```diff
@@ -259,9 +273,12 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<Set<VariableReferenc
             PlanNode rewrittenLeft = context.rewrite(joinNode.getLeft(), context.get());
             PlanNode rewrittenRight = context.rewrite(joinNode.getRight(), context.get());
 
+            PlanNodeStatsEstimate joinEstimate = statsProvider.getStats(joinNode);
+            boolean enabledByCostModel = strategy.equals(COST_BASED) && joinEstimate.getNullJoinBuildKeyCount() > getRandomizeOuterJoinNullKeyNullCountThreshold(session)
+                    && joinEstimate.getNullJoinBuildKeyCount() / joinEstimate.getJoinBuildKeyCount() > getRandomizeOuterJoinNullKeyNullRatioThreshold(session);
             List<JoinNode.EquiJoinClause> candidateEquiJoinClauses = joinNode.getCriteria().stream()
                     .filter(x -> isSupportedType(x.getLeft()) && isSupportedType(x.getRight()))
-                    .filter(x -> strategy.equals(ALWAYS) || (strategy.equals(KEY_FROM_OUTER_JOIN) && (context.get().contains(x.getLeft()) || context.get().contains(x.getRight()))))
+                    .filter(x -> enabledByCostModel || strategy.equals(ALWAYS) || (strategy.equals(KEY_FROM_OUTER_JOIN) && (context.get().contains(x.getLeft()) || context.get().contains(x.getRight()))))
```

Comment:
the strategy is fixed and doesn't need to get evaluated for each element, let's have a boolean variable outside the loop:
```
boolean isEnabled = enabledByCostModel || strategy.equals(ALWAYS) || (strategy.equals(KEY_FROM_OUTER_JOIN);
```

Reply:
(strategy.equals(KEY_FROM_OUTER_JOIN) is combined with other condition check involving the lambda element. We can combine "enabledByCostModel || strategy.equals(ALWAYS)" but still leave the (strategy.equals(KEY_FROM_OUTER_JOIN) alone.

Reply:
oh right!
yes, let's combine the first two. Can also introduce a variable for isEnabledForOuterJoinKeys and/or introduce a function for the outerjoin key check. I think shortening the check will make it a bit more readable

Reply:
> yes, let's combine the first two

I think it's fine to have the three conditions in the filter? As the saving for not needed to evaluate for each element is negligible, and we currently have three settings to enable the optimization, having all three should be better for readability and not confusing people on wondering if two are special and need to be combined first.

> Can also introduce a variable for isEnabledForOuterJoinKeys and/or introduce a function for the outerjoin key check

Added a function for the outer join key case

Reply:
it's more about readability than overhead: if I see something computed inside the loop, I think it is dependent on the loop

Reply:
one solution would be to hide all of the checks inside a function that gets called for every element

Reply:
But having all three checks in one function and gets called for every element does not help readability imo? 

We have three cases where the optimization should be enabled 1) Users set the session to always enable, not related to join stats and not join key specific 2) Users set the session to be cost based, and cost model shows that it can be enabled, not join key specific 3) Users set the session to be enabled when join key is from the output of an outer join, is join key specific and need to check for every join key. These three checks takes different inputs to make decisions.

Puting all three checks within one function and gets called for every element will hide the fact that the first two decisions are join key independent and does not need to be called for every element. And also this function will mix the arguments needed for each case which is not a win?